### PR TITLE
Note about C++ extensions

### PR DESCRIPTION
**Pull request by goldsborough**
I wrote a longer note describing usage of C++/CUDA extensions using our new mechanism.
Comments and thoughts welcome. Also let me know if it's ok to leave the samples [under my GitHub](https://github.com/goldsborough/pytorch-cpp-extension), or whether there's a better place to put it (e.g. `pytorch/examples`).

@zdevito @apaszke @colesbury

### CODE REVIEW

**Comment by soumith** [comment]
created https://github.com/pytorch/extension-cpp and gave you write access

**Comment by goldsborough** [comment]
Thanks! I've populated it.

File: docs/source/notes/cpp-extensions.rst [hunk 1]
```diff
@@ -0,0 +1,1020 @@
...
+functions) as well as the usual expressiveness of Python. However, there may be
+times when your operation is better implemented in C++. For example, your code
+may need to be *really* fast because it is called very frequently in your model
+or is very expensive even for few calls. Another plausible reason is that it
+depends on or interacts with other C or C++ libraries. To address such cases,
+PyTorch provides a very easy way of writing custom *C++ extensions*.
+
+C++ extensions are a mechanism we have developed to allow users (you) to create
+PyTorch operators defined *out-of-source*, i.e. separate from the PyTorch
+backend. This approach is *different* from the way native PyTorch operations are
```
Comment (ezyang) [inline docs/source/notes/cpp-extensions.rst]: Nit: I'm not sure how important it is for the reader to know how native PyTorch ops are implemented, except maybe as an aside.

File: docs/source/notes/cpp-extensions.rst [hunk 2]
```diff
@@ -0,0 +1,1020 @@
...
+The rest of this note will walk through a practical example of writing and using
+a C++ (and CUDA) extension. If you are being chased or someone will fire you if
+you don't get that op done by the end of the day, you can skip this section and
+head straight to the implementation details in the next section.
+
+Let's say you've come up with a new kind of recurrent unit that you found to
+have superior properties compared to the state of the art. This recurrent unit
+is similar to an LSTM, but differs in that it lacks a *forget gate* and uses an
+*Exponential Linear Unit* (ELU) as its internal activation function. Because
+this unit never forgets, we'll call it *LLTM*, or *Long-Long-Term-Memory* unit.
```
Comment (ezyang) [inline docs/source/notes/cpp-extensions.rst]: loooool

**Comment by ezyang** [comment]
It's very tutorial-y, so I wonder if the tutorial itself shouldn't live in `extension-cpp` or not. Hmm.

**Comment by goldsborough** [comment]
Wellll it was kinda intended to be the main source of documentation for the extensions. We'll want some kind of official docs for them. [Related docs](https://www.tensorflow.org/extend/adding_an_op) seemed to follow a similar example-based style. In this case I also added the whole benchmarking story just to motivate why one would write a C++ extension.

**Comment by ngimel** [comment]
This looks great!
I have couple questions
1) cuda kernels operate on raw data pointers, which would be pretty annoying for non-contiguous tensors, as it would require recreating collapseDims, IndexToOffset and similar functions currently in ATen/cuda/detail/TensorInfo.cuh or its THC copy (ATen/src/THC/THCTensorInfo.cuh). What would be the best way of exposing TensorInfo for use in custom kernels and setup functions?
2) How would one go about creating an extension where tensor arguments are of the different type? E.g. it may be desirable to have outputs of some reductions and norms on half tensor to be a float tensor. AT_DISPATCH_FLOATING_TYPES does not handle this situation now, it would require two type arguments, say scalar_t and accScalar_t.

**Comment by goldsborough** [comment]
Thanks @ngimel 
1. `ATen/cuda/detail/TensorInfo.cuh` does get shipped with the pytorch install, so if I understand correctly you could just include it and use it?
2. You can see a good example for this [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Embedding.cu#L337)

**Comment by ngimel** [comment]
1. I don't see TensorInfo.cuh installed in the binary install (torch.h is there, in site-packages/torch, but TensorInfo.cuh is found only in the source tree)
```
root@18f70a1b119f:/workspace# find / -name torch.h
/opt/pytorch/pytorch/torch/torch.h
/opt/conda/envs/pytorch-py3.6/lib/python3.6/site-packages/torch/lib/include/torch/torch.h
root@18f70a1b119f:/workspace# find / -name TensorInfo.cuh
/opt/pytorch/pytorch/aten/src/ATen/cuda/detail/TensorInfo.cuh
```
so a user who has only binary install and tries to compile extension using TensorInfo won't be able to do so, unless extension itself provides a copy of TensorInfo. 
2. Great, should work for tensor arguments too!

**Comment by goldsborough** [comment]
Is your fork up-to-date? The commit that installs those headers landed 4 days ago: https://github.com/pytorch/pytorch/pull/5474

**Comment by ngimel** [comment]
It was older than that, I was not paying attention. Thanks!

**Comment by goldsborough** [comment]
I am moving this PR to https://github.com/pytorch/tutorials/pull/214

**Pull request closed without merging**